# Ex2 - Getting and Knowing your Data

This time we are going to pull data directly from the internet.
Special thanks to: https://github.com/justmarkham for sharing the dataset and materials.

### Step 1. Import the necessary libraries

In [46]:
import pandas as pd
import numpy as np


### Step 2. Import the dataset from this [address](https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv). 

In [47]:
# import the dataset from a tsv file from the web
url = 'https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv'
chipotle = pd.read_csv(url, sep='\t')


### Step 3. Assign it to a variable called chipo.

In [48]:
chipo = chipotle

### Step 4. See the first 10 entries

In [49]:
# View the first 10 entries of chipo
chipo.head(10)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   
5         3         1                           Chicken Bowl   
6         3         1                          Side of Chips   
7         4         1                          Steak Burrito   
8         4         1                       Steak Soft Tacos   
9         5         1                          Steak Burrito   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98   
6                                                NaN     $1.69   
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75   
8  [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...     $9.25   
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...     $9.25

### Step 5. What is the number of observations in the dataset?

In [50]:
# Solution 1
chipo.shape[0]


4622

In [51]:
# Solution 2
len(chipo.index)


4622

### Step 6. What is the number of columns in the dataset?

In [52]:
# What is the number of columns in the dataset?
chipo.shape[1]

5

### Step 7. Print the name of all the columns.

In [53]:
# Print the name of all the columns.
chipo.columns  


Index(['order_id', 'quantity', 'item_name', 'choice_description',
       'item_price'],
      dtype='object')

### Step 8. How is the dataset indexed?

In [54]:
chipo.index

RangeIndex(start=0, stop=4622, step=1)

The dataset indexed with integers ranging from 0 to 4622

### Step 9. Which was the most-ordered item? 

In [55]:
# Find the most-ordered item, i.e the item with the highest quantity ordered
c = chipo.groupby('item_name')
c = c.sum()
c = c.sort_values(['quantity'], ascending=False)
most_ordered = c.head(1)

/tmp/ipykernel_126558/4157087645.py:3: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  c = c.sum()


### Step 10. For the most-ordered item, how many items were ordered?

In [56]:
num_items = most_ordered['quantity'].values[0]
print('The most ordered item is', most_ordered.index.values[0], 'and it was ordered', num_items, 'times')

The most ordered item is Chicken Bowl and it was ordered 761 times


### Step 11. What was the most ordered item in the choice_description column?

In [57]:
# Find the most ordered item in the choice_description column
# Solution 1 - consider each choice description as a single item
c = chipo.groupby('choice_description').sum()
c = c.sort_values(['quantity'], ascending=False)
most_ordered = c.head(1)
num_items = most_ordered['quantity'].values[0]
print('The most ordered item is', most_ordered.index.values[0], 'and it was ordered', num_items, 'times')

# Solution 2 - consider each choice description as a list of items
# calculate the most frequent subitem in the choice_description column
# chipo['choice_description'] = chipo['choice_description'].fillna('')  # fill NaN values with empty string
# chipo['choice_description'] = chipo['choice_description'].apply(lambda x: x[1:-1])  # remove the brackets
# chipo['choice_description'] = chipo['choice_description'].str.split(', ')  # split the string into a list
# chipo = chipo.set_index(['item_name', 'choice_description'])['quantity'].unstack().fillna(0)
# chipo = chipo.stack().reset_index().rename(columns={0:'quantity'})
# chipo = chipo.groupby('choice_description').sum()
# chipo = chipo.sort_values(['quantity'], ascending=False)
# most_ordered = chipo.head(1)



The most ordered item is [Diet Coke] and it was ordered 159 times


/tmp/ipykernel_126558/2467268564.py:3: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  c = chipo.groupby('choice_description').sum()


### Step 12. How many items were orderd in total?

In [61]:
# Sum up the quantities of all the items in the dataset
total_items_orders = chipo['quantity'].sum()
print('The total number of items ordered is', total_items_orders)

The total number of items ordered is 4972


### Step 13. Turn the item price into a float

#### Step 13.a. Check the item price type

In [68]:
chipo['item_price'].head()
# check typeof each element in the item_price column
# type(chipo['item_price'][0])
# the item_price column is of type string and it suffixes the price with a dollar sign

str

#### Step 13.b. Create a lambda function and change the type of item price

In [69]:
# create a lambda function that omit the dollar sign and convert the string to a float
convert_to_float = lambda x: float(x[1:-1])
chipo['item_price'] = chipo['item_price'].apply(convert_to_float)
chipo['item_price'].head()

0     2.39
1     3.39
2     3.39
3     2.39
4    16.98
Name: item_price, dtype: float64

#### Step 13.c. Check the item price type

In [70]:
# what is the type of the item_price column now?
type(chipo['item_price'][0])
# np.dtype('float64')

numpy.float64

### Step 14. How much was the revenue for the period in the dataset?

In [74]:
# calculate the revenue for the period in the dataset
revenue = (chipo['quantity'] * chipo['item_price']).sum()
print('The revenue was: $', revenue)

The revenue was: $ 39237.02


### Step 15. How many orders were made in the period?

In [76]:
# calculate the sum of the orders in the dataset

orders = chipo['order_id'].value_counts().count()
print('The total number of orders is', orders)

The total number of orders is 1834


### Step 16. What is the average revenue amount per order?

In [85]:
# Solution 1

# calculate the average revenue per order
avg_revenue_per_order = revenue / orders
print('The average revenue per order is $', avg_revenue_per_order)


The average revenue per order is $ 21.39423118865867


In [89]:
# Solution 2

# calculate the average revenue per order
chipo['revenue'] = chipo['quantity'] * chipo['item_price']
orders_grouped = chipo.groupby(by=['order_id']).sum()
avg_revenue_per_order = orders_grouped.mean()['revenue']
print('The average revenue per order is $', avg_revenue_per_order)


The average revenue per order is $ 21.39423118865867


/tmp/ipykernel_126558/2605967532.py:5: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  orders_grouped = chipo.groupby(by=['order_id']).sum()


### Step 17. How many different items are sold?

In [90]:
# calculate the number of unique items sold
unique_items = chipo['item_name'].value_counts().count()
print('The number of unique items sold is', unique_items)

The number of unique items sold is 50
